src/mappings/poktroll/suppliers.ts

```diff
@@ -0,0 +1,257 @@
+  stringify,
+  unprocessedEventHandler,
+  unprocessedMsgHandler,
+} from "../utils";
+
+export async function handleSupplierStakeMsg(
+  msg: CosmosMessage<MsgStakeSupplier>,
+): Promise<void> {
+  await attemptHandling(msg, _handleSupplierStakeMsg, unprocessedMsgHandler);
+}
+
+export async function handleUnstakeSupplierMsg(
+  msg: CosmosMessage<MsgUnstakeSupplier>,
+): Promise<void> {
+  await attemptHandling(msg, _handleUnstakeSupplierMsg, unprocessedMsgHandler);
+}
+
+export async function handleSupplierUnbondingBeginEvent(
+  event: CosmosEvent,
+): Promise<void> {
+  await attemptHandling(event, _handleSupplierUnbondingBeginEvent, unprocessedEventHandler);
+}
+
+export async function handleSupplierUnbondingEndEvent(
+  event: CosmosEvent,
+): Promise<void> {
+  await attemptHandling(event, _handleSupplierUnbondingEndEvent, unprocessedEventHandler);
+}
+
+async function _handleSupplierStakeMsg(msg: CosmosMessage<MsgStakeSupplier>) {
+  logger.debug(`[handleSupplierStakeMsg] (msg.msg): ${stringify(msg.msg, undefined, 2)}`);
+
+  if (!msg.msg.decodedMsg.stake) {
+    return logger.error(`[handleSupplierStakeMsg] stake not provided in msg`);
+  }
+
+  const msgId = messageId(msg)
+
+  const supplierMsgStake = MsgStakeSupplierEntity.create({
+    id: msgId,
+    signerId: msg.msg.decodedMsg.signer,
+    supplierId: msg.msg.decodedMsg.operatorAddress,
+    ownerId: msg.msg.decodedMsg.ownerAddress,
+    stakeAmount: BigInt(msg.msg.decodedMsg.stake.amount),
+    stakeDenom: msg.msg.decodedMsg.stake.denom,
+    blockId: msg.block.block.id,
+    transactionId: msg.tx.hash,
+    messageId: msgId,
+  })
+
+  const supplier = Supplier.create({
+    id: msg.msg.decodedMsg.operatorAddress,
+    operatorId: msg.msg.decodedMsg.operatorAddress,
+    ownerId: msg.msg.decodedMsg.ownerAddress,
+    stakeAmount: BigInt(msg.msg.decodedMsg.stake.amount),
+    stakeDenom: msg.msg.decodedMsg.stake.denom,
+    status: StakeStatus.Staked,
+    unstakingEndHeight: undefined,
+    unstakingEndBlockId: undefined,
+    unstakingBeginBlockId: undefined,
+  })
+
+  const servicesId: Array<string> = []
+  const supplierMsgStakeServices: Array<MsgStakeSupplierServiceProps> = []
+  const newSupplierServices: Array<SupplierServiceConfigProps> = []
+
+  const operatorAddress = msg.msg.decodedMsg.operatorAddress
+
+  for (const {endpoints, revShare, serviceId} of msg.msg.decodedMsg.services) {
+    servicesId.push(serviceId)
+
+    const endpointsArr = endpoints.map((endpoint) => ({
+      url: endpoint.url,
+      rpcType: endpoint.rpcType,
+      configs: endpoint.configs,
+    }))
+
+    const revShareArr = revShare.map((revShare) => ({
+      address: revShare.address,
+      revSharePercentage: revShare.revSharePercentage,
+    }))
+
+    supplierMsgStakeServices.push({
+      id: getMsgStakeServiceId(msgId, serviceId),
+      serviceId,
+      stakeMsgId: msgId,
+      endpoints: endpointsArr,
+      revShare: revShareArr
+    })
+
+    newSupplierServices.push({
```

Comment:
Similar to the feedback about about existing suppliers, we may have "existing" service configs (where identity is correctly defined by your ID function). 

**Question**: Assuming that we **did not** create new `SupplierServiceConfigs` when a supplier updates them, if we query for a supplier **and its related `serviceConfig`s** using subquery's historical state feature, do the values of the returned service configs reflect their respective historical state as well?

A. **If this is the case**, then we need to query for existing `SupplierServiceConfigs` (using the current ID function) and update them when they change.
B. **If this is not the case**, then we need to change the ID function to take the current height as a param to allow two configs for the same operator and service but different values to co-exist as distinct records.

Reply:
![image](https://github.com/user-attachments/assets/aca43d7c-946a-431e-a369-e6becd90816a)

Reply:
In relation to everything related to supplier services:

- I am currently using the `MsgStakeSupplierService` entity to save the configuration of the services that come from the messages.
- I am using the `SupplierServiceConfig` entity to have, at present, the suppliers' services, that is why I am removing the previous ones.

To search for the history of a supplier's services I thought you could use MsgStakeSupplierService. That's why I had both entities separately, one that was for history and the other for the present.

I understand that we were not understanding each other regarding these entities. I await your feedback to know how I proceed with them.

Reply:
Thank you for the detailed explanation! :raised_hands: 

I think I understand now. In order to use the historical state feature, we need to (soft) delete the existing entity and therefore, create a new one.